lib/addressable/uri.rb

```diff
@@ -1216,7 +1216,34 @@ def tld=(new_tld)
     # @example
     #   Addressable::URI.parse("http://www.example.co.uk").domain # => "example.co.uk"
     def domain
-      PublicSuffix.domain(self.host, ignore_private: true)
+      public_suffix.domain
+    end
+
+    ##
+    # Returns the first subdomain for this host.
+    #
+    # @example
+    #   Addressable::URI.parse("http://www.test.example.co.uk").domain # => "www"
+    def subdomain
+      subdomains.first
+    end
+
+    ##
+    # Returns the subdomains for this host.
+    #
+    # @example
+    #   Addressable::URI.parse("http://www.test.example.co.uk").domain # => ["www", "test"]
+    def subdomains
+      public_suffix.trd.to_s.split(".")
+    end
+
+    ##
+    # Returns a PublicSuffix::Domain object for this host.
+    #
+    # @example
+    #   Addressable::URI.parse("http://www.example.org").public_suffix #=> #<PublicSuffix::Domain @sld="example", @tld="org", @trd="www">
+    def public_suffix(**kwargs)
```

Comment:
I'm thinking it is better that users interact directly with `PublicSuffix` if they have the need to customise the options.

Reply:
Let's make this method private, use memoization (but cleared if `#host=` is used), and remove the arguments

Reply:
Done!